### PR TITLE
Router refactor

### PR DESCRIPTION
**Pull request by AntoineRR**
**Description**

This PR refactors the `routers` module of the Rust part of Robyn.
I noticed some improvements could be made there so here is my take on it.

Noticeable changes:
- Rename `Router` to `DynRouter`
- Introduce a `Router` trait to unify the behavior of `ConstRouter`, `DynRouter`, and `MiddlewareRouter`
- Get rid of the `get_relevant_map` methods (mapping is directly done at structure creation)
- Some parts of the code were made less verbose

I tried to keep the same logic as before. I think this could pave the way for further performance improvements for the router  :slightly_smiling_face:

### CODE REVIEW

**Comment by sansyrox** [comment]
@AntoineRR , great work. I will try to review this PR by today evening 😄

File: src/routers/const_router.rs
Comment (sansyrox) [inline src/routers/const_router.rs:26]: Hey @AntoineRR ,

I am bit unfamiliar with `traits` in rust. Can you please explain this statement to me. And the benefits associate with it?

File: src/routers/mod.rs
Comment (sansyrox) [inline src/routers/mod.rs:10]: what does the `(pub T)` mean here??

File: src/routers/mod.rs
Comment (sansyrox) [inline src/routers/mod.rs:26]: One nit: `get` function won't insert the function in the router.

File: src/routers/router.rs
Comment (sansyrox) [inline src/routers/router.rs:18]: Is `DynRouter` a `HTTPRouter`?

File: src/server.rs
Comment (sansyrox) [inline src/server.rs:392]: Do you have a better name suggestion for `tuple_params`?

**Review by sansyrox — COMMENTED** [review]
Thanks for the PR @AntoineRR :D . I have a few inline comments.

File: src/routers/const_router.rs
Comment (AntoineRR) [reply src/routers/const_router.rs:26]: Sure !

## About traits

Traits are a way in Rust to share behavior between structs. There is no inheritance in Rust as in a truly object oriented language like C++, so traits are more similar to interfaces in Java for example.

Here, defining the `Router` trait allows to specify what methods should be in a router : `add_route` and `get_route`. For every router we want to define, we have to implement this trait using the line above. This will force us to provide a definition for the `add_route` and `get_route` method that is specific to our struct, which is `ConstRouter` here.

## About the generics

I didn't want to change the code too much so I kept the same return types and parameter types in every router struct. The return type of the `get_route` method is different across the routers (`DynRouter` and `MiddlewareRouter` return an `Option<((PyFunction, u8), HashMap<String, String>)>` while `ConstRouter` return an `Option<String>`), so I used a generic type to unify their behavior in the trait. Similarly, the `route_type` parameter of the `get_route` method can either be a `MiddlewareRoute` or a `Method`. I needed two different generic types to handle those differences: this is `T` and `U` in the definition of the trait : `pub trait Router<T, U>`. For `ConstRouter`, `T` is `String` and `U` is `Method`.

## Benefits of using traits

With this PR, `DynRouter`, `ConstRouter` and `MiddlewareRouter` all implement the `Router` trait (I wanted to include the `WebSocketRouter` too but it seems too different from the others). This means the compiler is aware those share some behavior. We could now do some cool things like:
- Creating a `Vec<dyn Router>` containing the routers
- Create generic methods that take a type implementing the `Router` trait as parameter -> `fn method<T: Router>(router: T)`
- Eventually provide a default implementation for the `add_route` and `get_route` methods if we can make it the same for all structs (which I may try to do later)
- If a new method needs to be added for every router, its implementation can be in the `Router` trait instead of in each struct
- Maybe other things I forgot?

I hope this makes things clearer! In this PR I didn't take advantage of the trait for now, I wanted to do small iterations on the code instead of a really big one. I will try to improve things further if I have some time :wink:

File: src/routers/mod.rs
Comment (AntoineRR) [reply src/routers/mod.rs:10]: I used this struct as a wrapper for the `Method` and `MiddlewareRoute` types. RouteType is a struct that can contain any other type. `pub T` is an unnamed field of this struct, which is public and of type `T`. It can be accessed like this in an instance of type `RouteType`: `route_type.0`.

However now that I think about it it is probably unnecessary here, and I could just use a generic type instead of wrapping the parameter in this struct. I will push a new commit removing it soon

File: src/routers/mod.rs
Comment (AntoineRR) [reply src/routers/mod.rs:26]: Oops I copied the wrong comment, will fix that

File: src/routers/router.rs
Comment (AntoineRR) [reply src/routers/router.rs:18]: What do you mean by that?
This is the struct that was previously called `Router`. To my understanding it stores the routes that have parameters, and I called it `DynRouter` to oppose it to `ConstRouter`. I'm not familiar with routers so I may have not understood everything, let me know if you prefer me to rename it

File: src/server.rs
Comment (AntoineRR) [reply src/server.rs]: I'm not sure about it, but as I think the purpose of the middleware is to modify the request before it is routed, I called it `modified_request`?

File: src/routers/router.rs
Comment (sansyrox) [reply src/routers/router.rs:18]: @AntoineRR , the vanilla `Router`(now `DynRouter`) just serves HTTP Requests. I am not sure what you mean by `DynRouter`. I was thinking of something like `HTTPRouter` or `HTTPRequestRouter`.

What do you think?

Comment (AntoineRR) [reply src/routers/router.rs]: It makes sense! I renamed it to `HttpRouter` and changed the file name as well.

File: src/routers/mod.rs
Comment (sansyrox) [reply src/routers/mod.rs]: Ah, that makes sense. Thank you for explaining 😄

File: src/routers/const_router.rs
Comment (sansyrox) [reply src/routers/const_router.rs:25]: Thank you @AntoineRR . This an excellent explanation 😄

File: src/server.rs
Comment (sansyrox) [reply src/server.rs]: `modified_request` is much better than `tuple_params` Thanks :D

File: src/routers/const_router.rs
Comment (AntoineRR) [reply src/routers/const_router.rs:25]: You're welcome ! 😉

**Review by sansyrox — APPROVED** [review]
LGTM! 🚀  Great work. 

But I am holding this PR to allow (https://github.com/sansyrox/robyn/pull/297) to merge.

**Comment by AntoineRR** [comment]
Ok thanks! I'll solve the eventual merge conflicts once the other PR is merged

**Comment by sansyrox** [comment]
Looks like the other PR is gonna take some time. 

Great job on this one!

**Pull request merged**
